### ISSUE TITLE
Build failure: python312Packages.pytest-mock

### ISSUE THREAD

**Issue by n8henrie** [NixOS/nixpkgs#277591]
### Steps To Reproduce

Works for 311, but on 312 I get:

```console
$ nix build github:nixos/nixpkgs#python312Packages.pytest-mock
```

### Build log

```
tests/test_pytest_mock.py:800: could not import 'mock': No module named 'mock'
```

https://gist.github.com/n8henrie/3dca01475623b5b179e20b4eb216784b

### Additional context

Important dependent packages such as `python312Packages.mypy` and `python312Packages.tox` fail as a result (which is how I discovered this).

I'm surprised not to find an existing issue; if I've missed it please let me know.

### Notify maintainers

@dotlambda

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.8, NixOS, 23.11 (Tapir), 23.11.20231229.66559ca`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - channels(root): `""`
 - nixpkgs: `/nix/store/0y79nzj58rahr5z480xmxf7igxcb6x5n-source`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by dotlambda** [comment]
For me it works after 64eacbc2df9d9c5acae002de18dbf6a099d4a165 (on staging-next).

**Comment by n8henrie** [comment]
Thanks. Looks like a fix is merged in https://github.com/pytest-dev/pytest-mock/pull/403/commits/6da5b0506d6378a8dbe5ae314d5134e6868aeabd -- once this is released one could probably unhide the failing test.

**Comment by n8henrie** [comment]
I tested this morning on `master`, I wonder why a fix in `staging-next` didn't work? Testing again now.

**Comment by natsukium** [comment]
Python 3.12 support on master is not yet stable.
Many packages will be available with python3.12 when the current staging-next cycle is merged into master.
https://nixpk.gs/pr-tracker.html?pr=271586

**Comment by n8henrie** [comment]
> Python 3.12 support on master is not yet stable.

I'm not entirely sure what "stable" means in this context. Not yet stable as in "not in a stable release e.g. `release-23.11`"? Or just a general "not working quite right" status? I just noted that it was no longer listed as a python312.rcX and didn't see any warnings when I run

```console
$ nix run nixpkgs#python312 -- --version
Python 3.12.1
```

How can I tell whether a package in master is "stable"?

(Was good for me to read a bit more about the staging process: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#staging)

**Comment by dotlambda** [comment]
> How can I tell whether a package in master is "stable"?

According to the docs, Python packages are supported for 3.10 and 3.11. Once #275620 is merged it will be 3.11 and 3.12.

**Comment by natsukium** [comment]
Here is the section that explains this, but actually it would not be clear unless you are reasonably familiar with nixpkgs python.
I think it would be great to clarify the library support status along with the interpreter support at the top of the document.

https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#cpython-update-schedule-python-cpython-update-schedule

**Comment by SuperSandro2000** [comment]
The issue mentioned in the first comment is fixed on staging-next, so we can close this.
